internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,324 @@
+// pairing each proxy address with every target address. For each pair, it
+// generates a new [resolver.Address] using the proxy address, and adding the
+// target address as the attribute along with user info.
+func (r *delegatingResolver) combinedAddressesIfReadyLocked() ([]resolver.Address, []resolver.Endpoint, bool) {
+	if !r.targetResolverReady || !r.proxyResolverReady {
+		return nil, nil, false
+	}
+	var addresses []resolver.Address
+	for _, proxyAddr := range r.proxyAddrs {
+		for _, targetAddr := range r.targetResolverState.Addresses {
+			newAddr := resolver.Address{Addr: proxyAddr.Addr}
+			var user url.Userinfo
+			if r.proxyURL.User != nil {
+				user = *r.proxyURL.User
+			}
+			newAddr = proxyattributes.Populate(newAddr, proxyattributes.Options{
+				User:        user,
+				ConnectAddr: targetAddr.Addr,
+			})
+			addresses = append(addresses, newAddr)
+		}
+	}
+
+	// Create a list of combined addresses by pairing each proxy endpoint
+	// with every target endpoint. For each pair, it constructs a new
+	// [resolver.Endpoint] using the first address from the proxy endpoint,
+	// as the DNS resolver is expected to return only one address per proxy
+	// endpoint. The target address and user information from the proxy URL
+	// are added as attributes to this address. The resulting list of addresses
+	// is then grouped into endpoints, covering all combinations of proxy and
+	// target endpoints.
+	var endpoints []resolver.Endpoint
+	for _, endpt := range r.targetResolverState.Endpoints {
+		var addrs []resolver.Address
+		for _, proxyAddr := range r.proxyAddrs {
+			for _, targetAddr := range endpt.Addresses {
+				newAddr := resolver.Address{Addr: proxyAddr.Addr}
+				var user url.Userinfo
+				if r.proxyURL.User != nil {
+					user = *r.proxyURL.User
+				}
+				newAddr = proxyattributes.Populate(newAddr, proxyattributes.Options{
+					User:        user,
+					ConnectAddr: targetAddr.Addr,
+				})
+				addrs = append(addrs, newAddr)
+			}
+		}
+		endpoints = append(endpoints, resolver.Endpoint{Addresses: addrs})
+	}
+	return addresses, endpoints, true
+}
+
+// updateProxyResolverState updates the proxy resolver state by storing proxy
+// addresses and endpoints, marking the resolver as ready, and triggering a
+// state update if both proxy and target resolvers are ready. If the ClientConn
+// returns a non-nil error, it calls `ResolveNow()` on the target resolver.  It
+// is a StateListener function of wrappingClientConn passed to the proxy resolver.
+func (r *delegatingResolver) updateProxyResolverState(state resolver.State) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if logger.V(2) {
+		logger.Infof("Addresses received from proxy resolver: %s", state.Addresses)
+	}
+	r.proxyAddrs = state.Addresses
+	if r.proxyAddrs == nil && state.Endpoints != nil {
+		for _, endpoint := range state.Endpoints {
+			r.proxyAddrs = append(r.proxyAddrs, endpoint.Addresses...)
+		}
+	}
```

Comment:
You can probably write a helper or two to do the conversion from endpoints to addresses and vice versa based on which ones you need in both the state callbacks.

Reply:
We are only converting endpoints to addresses(or vice versa) in the case where the proxy address list is empty and we have endpoint list. Because we have to anyway flatten the endpoint list to combine with target endpoints and proxy resolver being a DNS resolver , address list and endpoints will have exact same addresses. And as for the target resolver, we never convert address to endpoint or vice versa, we combine based on what we get from target resolver and pass on to the ClientConn and let resolver_wrapper take care of the rest.

And since we have only one place, I remember a comment that is a function is called just once, we should rather have it inline if I remember correctly.